spyder/config/user.py

```diff
@@ -1,373 +1,456 @@
+
+    Notes
+    -----
+    The 'get' and 'set' arguments number and type differ from the overriden
+    methods.
     """
     DEFAULT_SECTION_NAME = 'main'
-    def __init__(self, name, defaults=None, load=True, version=None,
-                 subfolder=None, backup=False, raw_mode=False,
-                 remove_obsolete=False):
-        DefaultsConfig.__init__(self, name, subfolder)
-        self.raw = 1 if raw_mode else 0
-        if (version is not None and
-                re.match(r'^(\d+).(\d+).(\d+)$', version) is None):
-            raise ValueError("Version number %r is incorrect - must be in X.Y.Z format" % version)
-        if isinstance(defaults, dict):
-            defaults = [ (self.DEFAULT_SECTION_NAME, defaults) ]
-        self.defaults = defaults
-        if defaults is not None:
-            self.reset_to_defaults(save=False)
-        fname = self.filename()
-        if backup:
-            try:
-                shutil.copyfile(fname, "%s.bak" % fname)
-            except IOError:
-                pass
+
+    def __init__(self, name, path, defaults=None, load=True, version=None,
+                 backup=False, raw_mode=False, remove_obsolete=False):
+        """UserConfig class, based on ConfigParser."""
+        super(UserConfig, self).__init__(name=name, path=path)
+
+        self._load = load
+        self._version = self._check_version(version)
+        self._backup = backup
+        self._raw = 1 if raw_mode else 0
+        self._remove_obsolete = remove_obsolete
+        self._module_source_path = get_module_source_path('spyder')
+
+        self._defaults_folder = 'defaults'
+        self._backup_folder = 'backups'
+        self._backup_suffix = '.bak'
+        self._defaults_name_prefix = 'defaults'
+
+        # This is overriding a method from cp.ConfigParser
+        self.defaults = self._check_defaults(defaults)
+
+        fpath = self.get_config_path()
         if load:
             # If config file already exists, it overrides Default options:
-            self.load_from_ini()
+            self._load_from_ini()
             old_ver = self.get_version(version)
             _major = lambda _t: _t[:_t.find('.')]
             _minor = lambda _t: _t[:_t.rfind('.')]
+
             # Save new defaults
-            self._save_new_defaults(defaults, version, subfolder)
+            self._save_new_defaults(defaults, version, path)
+
             # Updating defaults only if major/minor version is different
             if _minor(version) != _minor(old_ver):
                 if backup:
                     try:
-                        shutil.copyfile(fname, "%s-%s.bak" % (fname, old_ver))
+                        shutil.copyfile(fpath, "{}-{}.bak".format(fpath,
+                                                                  old_ver))
                     except IOError:
                         pass
+
+                # TODO: This logic needs to be implemented outside as this is

```

Comment:
Since this class will be used by ay other third party plugin this logic cannot live here. However we should probably allow extensions so that this can be made. Third party plugin devs might want to also apply patches.

This does not need to be fix on this PR but I do want to think about options. 

One way would be to define a `UserConfigPatches` class that needs to define a couple of methods that can be registered with a configuration.

We should probably defer the `load` to user and not directly on the `__init__`